packages/reactivity/src/baseHandlers.ts

```diff
@@ -168,15 +168,16 @@ class MutableReactiveHandler extends BaseReactiveHandler {
     receiver: object
   ): boolean {
     let oldValue = (target as any)[key]
-    if (isReadonly(oldValue) && isRef(oldValue) && !isRef(value)) {
+    const isUnwrappingRef = isRef(oldValue) && !isRef(value)
```

Comment:
The old code was somewhat intentional, because the two `isRef` checks are not necessarily in many code paths. Ideally we should only perform the check in non-shallow && non-array cases, so this patch will introduce unwanted overhead for the other cases.

However this condition can indeed be optimized - I've done that in ef432bda7 and referenced this PR as a source of inspriation.